docs/tutorials/fourier_checking.ipynb

```diff
@@ -0,0 +1,1187 @@
+    "        accept_minus_reject_count += 1 if decision == \"accept\" else -1\n",
+    "    final_decision = \"accept\" if accept_minus_reject_count > 0 else \"reject\"\n",
+    "    res = res.append(\n",
+    "        {\n",
+    "            \"Distribution\": dist,\n",
+    "            \"Decision\": final_decision,\n",
+    "            \"Count\": 1\n",
+    "        }, ignore_index=True)\n",
+    "confusion = res.pivot_table(index=\"Distribution\",\n",
+    "                            columns=\"Decision\",\n",
+    "                            values=\"Count\",\n",
+    "                            aggfunc=\"sum\")\n",
+    "# Translate the counts into percentage\n",
+    "confusion.div(confusion.sum(axis=1), axis=0).apply(lambda x: round(x, 4) * 100)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "Tda2vy2pb1aU"
+   },
+   "source": [
+    "If you try the algorithm with more repetitions, you will observe the error rate decreases rapidly. You can also try different values of $K, N, $ and/or repetitions to see how the confusion matrix changes according.\n",
+    "\n",
+    "After you understand the concept of BPP, it is easy to understand [bounded-error quantum polynomial time (BQP)](https://en.wikipedia.org/wiki/BQP) now. BQP is the class of decision problems solvable by a quantum computer in polynomial time, with an error probability of at most 1/3 for all instances. It is the quantum analogue to the complexity class BPP. Actually, under quantum computer, previous problem can be solved through [Deutsch–Jozsa algorithm](https://en.wikipedia.org/wiki/Deutsch%E2%80%93Jozsa_algorithm) with a single query while guaranteed to be correct always. The Fourier Checking problem that will be introduced later is belong to BQP as well."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "spyqzGyf5kGW"
+   },
+   "source": [
+    "## Fourier Transform over $\\mathbb{Z}^n_2$\n",
+    "In this colab, we are interested in the boolean function of the form $f : \\{0, 1\\}^n \\rightarrow\\{-1, 1\\}$. In this case, the Fourier transform of $f$ over $\\mathbb{Z}^n_2$ is defined as\n",
+    "$$\n",
+    "  \\hat{f}(y) := \\frac{1}{\\sqrt{N}} \\sum_{x\\in\\{0,1\\}^n} (-1)^{x \\cdot y} f\n",
+    "  (x).\n",
+    "$$\n",
+    "where $x \\cdot y$ means the bit-wise inner product between $x$ and $y$. Note this is not the standard discrete Fourier transform definition over $\\mathbb{Z}_N$. According to the Parseval's identity, we have\n",
+    "\n",
+    "$$\n",
+    "  \\sum_{x\\in\\{0,1\\}^n} f(x)^2 = \\sum_{y\\in\\{0,1\\}^n} \\hat{f}(y)^2  = N .\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "id": "ucEcJmSQ26vg"
+   },
+   "outputs": [],
+   "source": [
+    "def bitwise_dot(x, y):\n",
+    "    \"\"\"Compute the dot product of two integers bitwise.\"\"\"\n",
+    "\n",
+    "    # fast bit parity\n",
+    "    def parallel_swar(i):\n",
```

Comment:
This is from the blogpost http://p-nand-q.com/python/algorithms/math/bit-parity.html, forgot to mention that. Now that I'm thinking about it I'd prefer rewriting it to a maybe slower but our own method. If we end up sticking with this version, we'll need to provide attribution, run a round with legal to see how we should do that.

Reply:
Agree. Change it into ```bin(i).count("1")```, which is much easier to read.

Reply:
Great, I would still call it bit parity instead of count_1s_in_z2...

Reply:
Done